tests/units/units.cpp

```diff
@@ -345,6 +345,173 @@ TEST(Units, replaceUnits)
     EXPECT_TRUE(m->replaceUnits(u2, u1));
 }
 
+TEST(Units, hasUnlinkedUnitsWhenBaseUnitsAddedToVariableButNotModel)
+{
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+
+    u1->setName("a_unit");
+
+    v1->setUnits(u1);
+    c1->addVariable(v1);
+    m->addComponent(c1);
+
+    EXPECT_TRUE(m->hasUnlinkedUnits());
+}
+
+TEST(Units, hasUnlinkedUnitsWhenBaseUnitsAddedToBothModelAndVariable)
+{
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+
+    u1->setName("a_unit");
+
+    v1->setUnits(u1);
+
+    c1->addVariable(v1);
+    m->addComponent(c1);
+    m->addUnits(u1);
+
+    EXPECT_FALSE(m->hasUnlinkedUnits());
+}
+
+TEST(Units, hasUnlinkedUnitsWhenBaseUnitsAddedToModelBeforeNameUsedForVariableUnits)
+{
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+
+    u1->setName("a_unit");
+
+    c1->addVariable(v1);
+    m->addComponent(c1);
+
+    // Adding units to model before adding to a variable
+    // means the units are linked.
+    m->addUnits(u1);
+    v1->setUnits("a_unit");
+
+    EXPECT_FALSE(m->hasUnlinkedUnits());
+}
+
+TEST(Units, hasUnlinkedUnitsWhenBaseUnitsAddedToModelAfterNameUsedForVariableUnits)
+{
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+
+    u1->setName("a_unit");
+
+    c1->addVariable(v1);
+    m->addComponent(c1);
+
+    // Adding units to model after adding to a variable
+    // means the units are not linked.
+    v1->setUnits("a_unit");
+    m->addUnits(u1);
+
+    EXPECT_TRUE(m->hasUnlinkedUnits());
+}
+
+TEST(Units, hasUnlinkedUnitsWhenNonBaseUnitsAddedToModelAfterNameUsedForVariableUnits)
+{
+    libcellml::ModelPtr m = libcellml::Model::create();
+    libcellml::ComponentPtr c1 = libcellml::Component::create();
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    libcellml::VariablePtr v1 = libcellml::Variable::create();
+
+    u1->setName("a_unit");
+    u1->addUnit("second");
+
+    c1->addVariable(v1);
+    m->addComponent(c1);
+
+    // Adding units to model after adding to a variable
+    // means the units are not linked.
+    v1->setUnits("a_unit");
+    m->addUnits(u1);
+
+    EXPECT_TRUE(m->hasUnlinkedUnits());
+}
```

Comment:
Not sure about the need of this test. It's the same as `hasUnlinkedUnitsWhenBaseUnitsAddedToModelAfterNameUsedForVariableUnits` except that we add `second` to `a_unit`, but why should that make things different when it comes to `hasUnlinkedUnits()`? What am I missing?...

Reply:
Isn't it to test that there is no special treatment of base units that is affecting this behaviour? (as we've found out in other aspects of units)

Reply:
Hmm... not convinced. We have `v1->setUnits("a_unit")`, which in itself results in an unlinked unit since `u1` is only added to the model after setting the unist to `v1`. Anyway, I am not going to have this hold back the PR. I just thought I would mention my "concerns" so that @hsorby can look into it when he's back.

Right now, the only thing that I want before accepting this PR is to fix the indentation of the math, i.e. https://github.com/cellml/libcellml/pull/483#discussion_r363150645.